user-auth/enclave/src/lib.rs

```diff
@@ -0,0 +1,161 @@
+#![no_std]
+#![deny(unsafe_op_in_unsafe_fn)]
+
+use std::slice;
+
+use argon2::{Algorithm, Argon2, Params, PasswordHasher, PasswordVerifier, Version};
+use once_cell::unsync::OnceCell;
+use password_hash::PasswordHashString;
+use sgx_tkeys::{SgxKeyPolicy, SgxSecret, SgxSecretBuilder};
+use sgx_tstd as std;
+use sgx_types::*;
+
+/// Length of a PHC-encoding of a 16 bit salt
+pub const SALT_LENGTH: usize = 22;
+
+/// It is assumed that the hash parameters and other configuration options (e.g.
+/// output length) are all set to their default values.  If any of these are
+/// changed in the future then this value will either need to be redetermined or
+/// dynammically calculated!
+// XXX: Certain assumptions are made here that relate to the hash context. See
+// the doc comment above.
+pub const HASH_STRING_LENGTH: usize = 97;
+
+/// Argon2 memory cost in MiB.
+pub const M_COST_MIB: u32 = 19;
+
+// Convert the Argon2 memory cost from MiB to KiB
+const fn m_cost_calc(mebibytes: u32) -> {
+    mebibytes * 1024
+}
+
+fn new_pepper(user_id: &[u8]) -> SgxSecret {
+    SgxSecretBuilder::new()
+        .key_id(&user_id)
+        .policy(SgxKeyPolicy::MRSIGNER)
+        .build()
+}
+fn new_hash_context<'a>(pepper: &'a [u8]) -> Result<Argon2<'a>, argon2::Error> {
+    Argon2::<'a>::new_with_secret(
+        pepper.as_ref(),
+        Algorithm::default(),
+        Version::default(),
+        // The parameters chosen here are, respectively, the maximum between the
+        // default set by the Argon2 library and the value recommended at:
+        //
+        // https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html
+        Params::new(m_cost_calc(M_COST_MIB), 3, 1, None).unwrap(),
+    )
+}
+
+///# Safety: guaranteed by the trusted bridge routine generated with the
+/// Intel SDK's `sgx_edger8r` tool
+#[no_mangle]
+pub unsafe extern "C" fn hash_password(
+    // Input paramaters
+    password: *const u8,
+    password_length: usize,
+    salt: *const u8,
+    user_id: *const u8,
+    user_id_length: usize,
+    // Output paramaters
+    //
+    // See https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md#argon2-encoding
+    hash_string: *mut u8,
+) -> sgx_status_t {
+    let password_buf = OnceCell::<&[u8]>::new();
+    let user_id_buf = OnceCell::<&[u8]>::new();
+    let salt_buf = OnceCell::<&str>::new();
+    unsafe {
+        password_buf
+            .set(slice::from_raw_parts(password, password_length))
+            .unwrap();
+        user_id_buf
+            .set(slice::from_raw_parts(user_id, user_id_length))
+            .unwrap();
+        salt_buf
+            .set(std::str::from_utf8(slice::from_raw_parts(salt, SALT_LENGTH)).unwrap())
+            .unwrap();
+    }
+
+    let pepper = new_pepper(user_id_buf.get().unwrap());
+    let hash_ctx = new_hash_context(pepper.as_ref());
+
+    match hash_ctx {
+        Err(_) => sgx_status_t::SGX_ERROR_UNEXPECTED,
```

Comment:
would be kool to display the error value (perhaps with `eprintln!` OCALL), of if it really is not useful, use an `if let Ok(ctx)` to reduce right-ward drift (and perhaps improve readability)